compiler/rustc_mir_build/src/thir/pattern/usefulness.rs

```diff
@@ -1075,24 +1085,44 @@ impl<'tcx> Witness<'tcx> {
     }
 }
 
-/// Report that a match of a non_exhaustive type marked with `reachable_patterns` is not
-/// exhaustive enough.
+fn if_let_span(tcx: TyCtxt<'_>, pat_id: HirId) -> Option<Span> {
+    let hir = tcx.hir();
+    let parent = hir.get_parent_node(pat_id);
+    let parent_parent = hir.get_parent_node(parent);
+    let parent_parent_node = hir.get(parent_parent);
+
+    if let hir::Node::Expr(hir::Expr {
+        kind: hir::ExprKind::If(hir::Expr { kind: hir::ExprKind::Let(pat, ..), .. }, ..),
+        ..
+    }) = parent_parent_node
+    {
+        Some(pat.span)
+    } else {
+        None
+    }
+}
```

Comment:
This looks dodgy, we need a more robust way get that info. First in `check_match` there's a `source` variable that can be `hir::MatchSource::IfLetDesugar`, which will tell us if the match came from an `if let`. Then if we want the pattern we can find it in the match arms.
What happens if you keep `pcx.span`? Does it point to the wrong thing? Otherwise I'd want to change the span of the wildcard arm in `check_match`

Reply:
Yeah, the `pcx.span` points to nothing so we get the weird error message that has no span and so no content.

Is the `MatchSource::IfLetDesugar` something that recently changed, currently there is no `if let ..` variant I think the lint is being triggered by [check_let](https://github.com/rust-lang/rust/blob/master/compiler/rustc_mir_build/src/thir/pattern/check_match.rs#L148) and [check_let_reachablility](https://github.com/rust-lang/rust/blob/master/compiler/rustc_mir_build/src/thir/pattern/check_match.rs#L427).

I just copied this from https://github.com/rust-lang/rust/blob/master/compiler/rustc_mir_build/src/thir/pattern/check_match.rs#L774. I agree it feels a bit wonky but since the `MatchSouce` doesn't have any info about `IfLetDesugar` and we are only called when the expression is an `if let ..` are there other options?

Reply:
The other way to handle this would be to change the `Usefulness::is_useful` method to
```rust
fn is_useful(&self) -> bool {
    match self {
        Usefulness::NoWitnesses(set) => !set.is_empty(),
        Usefulness::WithWitnesses(_) => false,
    }
}
```

Then because no `Usefulness::WithWitness` would trigger the lint there are no wildcards that are created from
```rust
let wild_pattern = cx.pattern_arena.alloc(Pat::wildcard_from_ty(scrut_ty));
let v = PatStack::from_pattern(wild_pattern);
```

I think this might actually be the reason for the dummy span, not the fact that it's a desugaring expansion (since it isn't, I was wrong about that). I'm pretty sure it's this https://github.com/rust-lang/rust/blob/master/compiler/rustc_mir_build/src/thir/pattern/usefulness.rs#L1233-L1235

Reply:
Ohh you're right! So indeed `MatchSource::IfLetDesugar` has disappeared, I was on an old commit. And also we get the dummy span because we're in `WithWitnesses`/exhaustiveness mode. I was sure that `if let` added an extra wildcard arm themselves, but that's because [this line](https://github.com/rust-lang/rust/blob/1cd17addad57e2aaf5dc6d3febe042ebb733bee3/compiler/rustc_mir_build/src/thir/pattern/check_match.rs#L443) believes it too. But that line is actually unreachable, I just tested it x).
Ok so you're right, a good solution would be to never lint when we're in `WithWitnesses`/exhaustiveness mode. And it does make sense not to lint on `if let`; no one expects an exhaustive list of patterns there.

Reply:
The logic you suggest is good. If you'll let me nitpick a bit: to make the intent clear I'd rather we check `usefulness.is_useful() && matches!(witness_preference, LeaveOutWitness)` with a comment that explains that we only want to lint on a wildcard written by the user.
That makes me think we could rename `WitnessPreference::LeaveOutWitness/ConstructWitness` to like `ArmType::RealArm/FakeExtraWildcardArm` ^^